components/eventing-controller/controllers/subscriptionv2/jetstream/reconciler.go

```diff
@@ -197,81 +184,121 @@ func (r *Reconciler) enqueueReconciliationForSubscriptions(subs []eventingv1alph
 		if err := r.Backend.DeleteSubscription(subscription); err != nil {
 			// if failed to delete the external dependency here, return with error
 			// so that it can be retried
-			return pkgerrors.MakeError(errFailedToDeleteSub, err)
+			if syncErr := r.syncSubscriptionStatus(ctx, subscription, err, log); syncErr != nil {
+				return ctrl.Result{}, syncErr
+			}
+			return ctrl.Result{}, pkgerrors.MakeError(errFailedToDeleteSub, err)
 		}
 
-		// remove our finalizer from the list and update it.
+		// remove the eventing finalizer from the list and update the subscription.
 		subscription.ObjectMeta.Finalizers = utils.RemoveString(subscription.ObjectMeta.Finalizers,
 			eventingv1alpha2.Finalizer)
-		if err := r.Client.Update(ctx, subscription); err != nil {
-			events.Warn(r.recorder, subscription, events.ReasonUpdateFailed, "Update Subscription failed %s", subscription.Name)
-			return pkgerrors.MakeError(errFailedToRemoveFinalizer, err)
+		return ctrl.Result{}, r.syncSubscriptionStatus(ctx, subscription, nil, log)
+	}
+	return ctrl.Result{}, nil
+}
+
+// syncSubscriptionStatus syncs Subscription status and updates the k8s subscription.
+func (r *Reconciler) syncSubscriptionStatus(ctx context.Context,
+	desiredSubscription *eventingv1alpha2.Subscription, err error, log *zap.SugaredLogger) error {
+	// set ready state
+	desiredSubscription.Status.Ready = err == nil
+
+	// compile the desired conditions
+	desiredSubscription.Status.Conditions = eventingv1alpha2.GetSubscriptionActiveCondition(desiredSubscription, err)
+
+	// Update the subscription
+	return r.updateSubscription(ctx, desiredSubscription, log)
+}
+
+// updateSubscription updates the subscription changes to k8s.
+func (r *Reconciler) updateSubscription(ctx context.Context,
+	sub *eventingv1alpha2.Subscription, logger *zap.SugaredLogger) error {
+	namespacedName := &k8stypes.NamespacedName{
+		Name:      sub.Name,
+		Namespace: sub.Namespace,
+	}
+
+	// fetch the latest subscription object, to avoid k8s conflict errors
+	actualSubscription := &eventingv1alpha2.Subscription{}
+	if err := r.Client.Get(ctx, *namespacedName, actualSubscription); err != nil {
+		return err
+	}
+
+	// copy new changes to the latest object
+	desiredSubscription := actualSubscription.DeepCopy()
+	desiredSubscription.Status = sub.Status
+	desiredSubscription.ObjectMeta.Finalizers = sub.ObjectMeta.Finalizers
+
+	// sync subscription status with k8s
+	if err := r.updateStatus(ctx, actualSubscription, desiredSubscription, logger); err != nil {
+		r.namedLogger().Errorf("updateStatus() error: %v", err)
+		return pkgerrors.MakeError(errFailedToUpdateStatus, err)
+	}
+
+	// update the subscription object in k8s
+	if !reflect.DeepEqual(actualSubscription.ObjectMeta.Finalizers, desiredSubscription.ObjectMeta.Finalizers) {
+		if err := r.Update(ctx, desiredSubscription); err != nil {
+			r.namedLogger().Errorf("updateStatus() error: %v", err)
+			return pkgerrors.MakeError(errFailedToAddFinalizer, err)
```

Comment:
is this also the error that is also thrown when you want failed to remove finalizer? Isn't this conflicting information?

Reply:
I change the error message to a generic one (failed to remove/add)